File: statics/js/app/helper/gitlab.js

define([
	"app",
], function(app){
	var $http = app.ng_objects.$http;
	var config = app.objects.config;
	var gitlabMap = {};
	function getGitlab(name) {
		gitlabMap[name] = gitlabMap[name] || {
			inited: false,                                          // is already init
			username: '',   // gitlab 用户名                        // gitlab username
			lastCommitId: "master",                                // 最新commitId  替代master  用于cdn加速
			projectId: undefined,                                  // project id
			projectName: 'keepworkdatasource',                   // repository name
			projectMap:{},                                      // 项目列表
			apiBaseUrl: 'http://git.keepwork.com/api/v4',     // api base url
			rawBaseUrl: 'http://git.keepwork.com',              // raw base url
			rootPath: '',                                           // 根路径
			httpHeader: {},
		}

		return gitlabMap[name];
	}
	// http请求
	function httpRequest(self, method, url, data, success, error) {
		self.dataSource.dataSourceToken && (self.httpHeader["PRIVATE-TOKEN"] = self.dataSource.dataSourceToken);
		//console.log(url);
		var config = {
			method: method,
			url: self.apiBaseUrl + url,
			headers: self.httpHeader,
			skipAuthorization: true,  // 跳过插件satellizer认证
			isShowLoading:data.isShowLoading == undefined ? true : data.isShowLoading,
		};

		data = data || {};
		data.per_page = 100;

		if (method == "POST" || method == "PUT") {
			config.data = data;
		} else {
			config.params = data;
		}

		var result = undefined;
		var _success = function (response) {
			//console.log(response);
			if (response.status < 200 || response.status >=300) {
				error && error(response);
				return;
			}

			var headers = (typeof(response.headers) == "function") && response.headers();
			if (headers && headers["x-next-page"] && data.isFetchAll) {
				data.page = parseInt(headers["x-next-page"]);
				result = (result || []).concat(response.data);
				//console.log(result);
				$http(config).then(_success).catch(failed);
			} else {
				//console.log(response);
				result = result ? (result.concat(response.data)) : response.data;
				typeof success == 'function' && success(result);
			}
		};
		var failed = function (response) {
			//console.log(response);
			typeof error == 'function' && error(response);
		};

		$http(config).then(success).catch(failed);
	}

	function createWebhook(self) {
		var hookUrl = config.apiUrlPrefix + "data_source/gitlab_webhook";
		//var hookUrl = "http://dev.keepwork.com/api/wiki/models/data_source/gitlabWebhook";
		var isExist = false;
		httpRequest(self, "GET", "/projects/" + self.projectId + "/hooks", {}, function (data) {
			for (var i = 0; i < data.length; i++) {
				if (data[i].url == hookUrl && data[i].push_events) {
					isExist = true;
				}
			}
			// 不存在创建
			if (!isExist) {
				httpRequest(self, "POST", "/projects/" + self.projectId + "/hooks", {
					url: hookUrl,
					push_events: true,
					enable_ssl_verification: false,
				}, function () {
					console.log("webhook create success");
				}, function () {
					console.log("webhook create failed");
				});
			}
		}, function () {

		});
	}

	function listCommits(self, data, success, error) {
		var url = '/projects/' + self.projectId + '/repository/commits';
		httpRequest(self, 'GET', url, data, cb, errcb);
	}

	function getLastCommitId(self, success, error, isShowLoading) {
		listCommits(self, {isShowLoading:isShowLoading}, function (data) {
			if (data && data.length > 0) {
				self.lastCommitId = data[0].id;
			} else {
				self.lastCommitId = "master";
			}
			success && success(self.lastCommitId);
		}, error);
	}

	function createProject(self, success, error) {
		var projectName = self.projectName;
		var visibility = self.visibility;
		var _success = function(params) {
			self.projectId = params.projectId;
			createWebhook(self);
			// 更新项目ID
			//util.post(config.apiUrlPrefix + 'site_data_source/updateById', {_id:self.dataSource.data_source_id, project_id:params.projectId, project_name:projectName});

			getLastCommitId(self, function(lastCommitId){
				self.lastCommitId = lastCommitId;
			});

			success && success();	
			return;
		}

		httpRequest(self, "GET", "/projects", {search: projectName, owned: true}, function (projectList) {
			var project = undefined;
			var method = "POST";
			var url = "/projects";
			var data = {name:projectName, visibility: visibility, request_access_enabled:true};


			// 查找项目是否存在
			for (var i = 0; i < projectList.length; i++) {
				if (projectList[i].name.toLowerCase() == projectName.toLowerCase()) {
					project = projectList[i];
					break;
				}
			}

			// 不存在或需要修改
			if (!project) {
				httpRequest(self, method, url, data, function (project) {
					if (project) {
						successCallback({projectId:project.id, projectName:params.projectName,lastCommitId:params.lastCommitId});
					} else {
						error && error();
					}
					//self.getLastCommitId(cb, errcb);
				}, errcb);
			} else if (project.visibility != visibility) {
				method = "PUT";
				url += "/" + project.id;
				data.id = project.id;
				
				// 不存在则创建项目 存在更新
				httpRequest(self, method, url, data, function (project) {
					_success({projectId:project.id, projectName:params.projectName,lastCommitId:params.lastCommitId});
				}, errcb);
			} else {
				_success({projectId:project.id, projectName:params.projectName,lastCommitId:params.lastCommitId});
			}
		}, error);
	}

	function init(self, dataSource, success, error) {
		if (self.inited) {
			success && success();
			return;
		}
		//console.log(dataSource);
		self.type = dataSource.type;
		self.username = dataSource.external_username;
		self.httpHeader["PRIVATE-TOKEN"] = dataSource.token;
		self.dataSourceToken = dataSource.token;
		self.apiBaseUrl = dataSource.api_base_url;
		self.rawBaseUrl = dataSource.raw_base_url;
		self.lastCommitId = dataSource.last_commit_id || "master";
		self.projectName = dataSource.project_name || self.projectName;
		self.projectId = dataSource.project_id || undefined;
		self.visibility = dataSource.visibility || "private";
		self.dataSource = dataSource;

		self.keepwrokUsername = dataSource.username;
		self.keepworkSitename = dataSource.sitename;

		if (!self.username || !self.dataSourceToken || !self.apiBaseUrl || !self.rawBaseUrl) {
			console.log("gitlab data source init failed!!!");
			error && error();
			return;
		}
		
		if (self.isInited || self.projectId) {
			//self.getLastCommitId(function(lastCommitId){
				//lastCommitId && (self.lastCommitId = lastCommitId);
				//cb && cb();
			//}, errcb);
			self.inited = true;
			success && success();
			return;
		}

		createProject(self, {projectName:self.projectName, visibility:self.visibility, lastCommitId:self.lastCommitId}, function() {
			self.inited = true;
			success && success();
		}, error);

		return;
	};


	function getFileUrlPrefix(self) {
		return '/projects/' + self.projectId + '/repository/files/';
	}

    function _encodeURIComponent(url) {
		return encodeURIComponent(url).replace(/\./g,'%2E')
    }
	function getCommitMessagePrefix() {
		return "keepwork commit: ";
	}

	return {
		registerDataSource: function(name, dataSource, success, error) {
			var inst = getGitlab(name);
			inst.getLastCommitId = function(success, error) {
				getLastCommitId(inst, success, error);
			}
			inst.setLastCommitId = function(lastCommitId) {
				inst.lastCommitId = lastCommitId;
			}
			inst.getTree = function(params, success, error) {
				var self = this;
				var url = '/projects/' + self.projectId + '/repository/tree';
				var path = params.path || "";
				params.path = path.substring(1);
				params.recursive = params.recursive == undefined ? true : params.recursive;
				params.isFetchAll = params.recursive;
				httpRequest(self, "GET", url, params, function (data) {
					var pagelist = [];
					for (var i = 0; i < data.length; i++) {
						var path = '/' + data[i].path;
						var page = {pagename: data[i].name};
						var suffixIndex = path.lastIndexOf(".md");
						// 不是md文件不编辑
						if (suffixIndex < 0)
							continue;

						page.url = path.substring(0, path.lastIndexOf('.'));
						var paths = page.url.split('/');
						if (paths.length < 3)
							continue;

						page.username = paths[1];
						page.sitename = paths[2];
						page.pagename = paths[paths.length - 1];
						page.blobId = data[i].id; // 文档sha

						pagelist.push(page);
					}
					success && success(pagelist);
				}, error);
			}
			inst.writeFile = function(params, success, error) {
				var self = this;
				params.path = params.path.substring(1);
				var url = getFileUrlPrefix(self) + _encodeURIComponent(params.path);
				params.commit_message = getCommitMessagePrefix() + params.path;
				params.branch = params.branch || "master";
				httpRequest(self, "GET", url, {path: params.path, ref: params.branch, isShowLoading:params.isShowLoading}, function (data) {
					// 已存在
					if (data && data.blob_id) {
						httpRequest(self, "PUT", url, params, success, error)
					} else {
						httpRequest(self, "POST", url, params, success, error);
					}
				}, function () {
					httpRequest(self, "POST", url, params, success, error);
				});
		
			}
			init(inst, dataSource, success, error);
		},
		getDataSource: function(name) {
			return getGitlab(name);
		},
	}
})

File: helper/gitlab.lua

local gitlab = {}
local l_admin_token = nws.config.gitlab.token
local l_gitlab_host = "http://" .. nws.config.gitlab.host
local l_default_project_name = "keepworkdatasource"
local l_api_base_url = l_gitlab_host .. "/api/v4"
local l_raw_base_url = l_gitlab_host 
local util = nws.util

function gitlab:init(config)
	self.api_base_url = config.api_base_url
	self.raw_base_url = config.raw_base_url
	self.token = config.token
	self.project_name = config.project_name or l_default_project_name
	self.project_id = config.project_id

	return self
end

function gitlab:new(config) 
	local obj = {}
	
	setmetatable(obj, self)
	self.__index = self

	obj:init(config)

	return obj
end

function gitlab:get_user() 
	local res = util.get_url({
		url = self.api_base_url .. "/user",
		method = "GET",
		headers = {['PRIVATE-TOKEN'] = self.token},
	})

	if not res or res.status_code ~= 200 then
		return "get gitlab user error"
	end

	return nil, res.data
end

-- 创建gitlab用户
function gitlab:create_user(params)
	if not params.username then
		return "用户名为空"
	end
	
	local gitlab_user_params = {}
	gitlab_user_params.username = params.username
	gitlab_user_params.name = gitlab_user_params.username
	gitlab_user_params.email = gitlab_user_params.name .. '@paraengine.com'
	gitlab_user_params.password = params.password or tostring(os.time())
	gitlab_user_params.skip_confirmation = true

	local res = util.get_url({
		url = self.api_base_url .. "/users",
		method = "GET",
		headers = {['PRIVATE-TOKEN']= self.token},
		data = {username=params.username},
	})

	if not res or res.status_code ~= 200 then
		return "get gitlab user failed"
	end
	
	if res.data and #res.data == 1 then
		return nil, res.data[1]
	end

	print("gitlab用户不存在，创建gitlab用户")
	-- 不存在创建用户
	res = util.get_url({
		url = api_base_url .. "/users",
		method = "POST",
		headers = {['PRIVATE-TOKEN']= self.token},
		data = gitlab_user_params,
	})
	if not res or res.status_code ~= 201 then
		return "create gitlab user failed"
	end

	return nil, res.data
end


-- 或取gitlab user认证token  不存在创建
function gitlab:get_user_token(params)
	if not params.user_id then
		return "user id is null"
	end
	local user_id = params.user_id

	local res = util.get_url({
		url=self.api_base_url .. "/users/" .. tostring(user_id) .. "/impersonation_tokens",
		method = "GET",
		headers = {['PRIVATE-TOKEN']= self.token},
	})

	if not res or res.status_code ~= 200 then
		return "get gitlab user token failed"
	end

	local tokens = res.data or {}
	for _, token in ipairs(tokens) do
		if not token.revoked and token.active then
			return nil, token	
		end
	end

	res = util.get_url({
		url=self.api_base_url .. "/users/" .. tostring(user_id) .. "/impersonation_tokens",
		method = "POST",
		headers = {['PRIVATE-TOKEN']= self.token},
		data = {
			["name"] = "keepwork", 
			["expires_at"] = "2222-12-12", 
			["scopes"]= {"api", "read_user"},
		},
	})
	if not res or res.status_code ~= 201 then
		return "create user token failed"
	end

	return nil, res.data
end

-- 创建项目
function gitlab:create_project(params)
	params.project_name = params.project_name or self.project_name
	if not params.project_name then
		return "project name is null"
	end
	params.visibility = params.visibility or "private"
	local res = util.get_url({
		url = self.api_base_url .. "/projects",
		method = "GET",
		headers = {['PRIVATE-TOKEN'] = self.token},
		data = {
			search=params.project_name,
			owned=true,
		},
	})

	if not res or res.status_code ~= 200 then
		return "获取项目信息失败"
	end
	
	data = res.data or {}

	for _, project in ipairs(data) do
		if string.lower(project.name) == string.lower(params.project_name) then
			return nil, project
		end
	end

	-- 不存创建项目
	res = util.get_url({
		url = self.api_base_url .. "/projects",
		method = "POST",
		headers = {['PRIVATE-TOKEN'] = self.token},
		data = {
			name = params.project_name, 
			request_access_enabled = true, 
			visibility = params.visibility,
		},
	})

	if not res or res.status_code ~= 201 then
		return "创建项目失败"
	end
	
	self.project_id = res.data.id

	self:write_file({path="README.md", content=" "})
	return nil, res.data
end

-- 写文件
function gitlab:write_file(params)
	params.project_id = params.project_id or self.project_id
	if not params.path or not params.content or not params.project_id then
		return "参数错误"
	end

	local method = "PUT"
	local path = util.encode_url(params.path)
	local arg_str = "?ref=" .. (params.ref or "master") 
	local res = util.get_url({
		url = self.api_base_url .. "/projects/" .. tostring(params.project_id) .. "/repository/files/" .. path .. arg_str,
		method = "GET",
		headers = {['PRIVATE-TOKEN'] = self.token},
	});

	if not res or res.status_code ~= 200 then
		--print("获取文件信息失败, 创建文件")
		method = "POST"
	end
	
	res = util.get_url({
		url = self.api_base_url .. "/projects/" .. tostring(params.project_id) .. "/repository/files/" .. path,
		method = method,
		headers = {['PRIVATE-TOKEN'] = self.token},
		data = {
			branch = params.branch or "master",
			encoding = params.encoding or "text",
			content = params.content,
			commit_message = "keepwork commit file: " .. params.path,
		}
	})
	if not res or (res.status_code ~= 200 and res.status_code ~= 201) then
		return "xie文件失败"
	end

	return res.data
end

-- 读文件
function gitlab:get_file(params)
	params.project_id = params.project_id or self.project_id
	if not params.path or not params.project_id then
		return "参数错误"
	end

	local path = util.encode_url(params.path)
	local res = util.get_url({
		url = self.api_base_url .. "/projects/" .. tostring(params.project_id) .. "/repository/files/" .. path .. "?ref=master",
		method = "GET",
		headers = {['PRIVATE-TOKEN'] = self.token},
	});

	if not res or res.status_code ~= 200 then
		return "获取文件内容失败"
	end

	return nil, res.data
end

-- 获取文件内容
function gitlab:get_content(params)
	local err, data = self:get_file(params)

	if not data then
		return err
	end

	return nil, util.decode_base64(data.content)
end


return gitlab







